

<h1 font size = 5>Lab 4 Từ Hiểu biết đến Chuẩn bị</font></h1>

Thời lượng: 120 phút

## Giới thiệu

Trong lab này, chúng ta sẽ tiếp tục tìm hiểu về phương pháp luận khoa học dữ liệu và tập trung vào các giai đoạn **Hiểu biết dữ liệu** và **Chuẩn bị dữ liệu**.

## Mục lục

<div class="alert alert-block alert-info" style="margin-top: 20px">

1. [Tóm tắt lại](#0) <br>
2. [Hiểu biết dữ liệu](#2) <br>
3. [Chuẩn bị dữ liệu](#4) <br>
</div>
<hr>

# Tóm tắt lại <a id="0"></a>

Trong bài Lab **Từ Yêu cầu đến Thu thập**, chúng ta biết rằng dữ liệu chúng ta cần để trả lời câu hỏi được phát triển trong giai đoạn Hiểu biết nghiệp vụ (cụ thể là *chúng ta có thể tự động hóa quy trình xác định nguồn gốc ẩm thực của một công thức nhất định không?*) đã sẵn sàng. Nhà nghiên cứu Yong-Yeol Ahn đã tìm kiếm hàng chục nghìn công thức nấu ăn (các món ăn và nguyên liệu) từ ba trang web khác nhau, đó là:

<img src = "https://ibm.box.com/shared/static/4fruwan7wmjov3gywiz3swlojw0srv54.png" width=500>

www.allrecipes.com

<img src = "https://ibm.box.com/shared/static/cebfdbr22fjxa47lltp0bs533r103g0z.png" width=500>

www.epicurious.com

<img src = "https://ibm.box.com/shared/static/epk727njg7xrz49pbkpkzd05cm5ywqmu.png" width=500>

www.menupan.com

Để biết thêm thông tin về Yong-Yeol Ahn và nghiên cứu của anh ấy, bạn có thể đọc bài báo của anh ấy trên [Flavor Network and the Principles of Food Pairing](http://yongyeol.com/papers/ahn-flavornet-2011.pdf).

Chúng tôi cũng đã thu thập dữ liệu và đặt nó trên máy chủ IBM để thuận tiện cho bạn.

------------

# Tìm hiểu dữ liệu <a id="2"></a>

<img src="https://ibm.box.com/shared/static/89geb3m0ge1z73s92hl8o8wdcpcrggtz.png" width=500>

<strong> Lưu ý quan trọng:</strong> Xin lưu ý rằng bạn không cần biết cách lập trình trong Python. Đoạn code sau đây nhằm minh họa các giai đoạn hiểu dữ liệu và chuẩn bị dữ liệu, vì vậy sẽ không có vấn đề gì nếu bạn chưa hiểu từng dòng code riêng lẻ. Chúng ta có một khóa học đầy đủ về lập trình bằng Python, <a href="http://cocl.us/PY0101EN_DS0103EN_LAB3_PYTHON_Coursera"> <strong> Python cho Khoa học dữ liệu </strong> </a>, cũng được cung cấp trên Coursera. Vì vậy, hãy hoàn thành khóa học Python nếu bạn muốn học cách lập trình bằng Python.

### Sử dụng notebook này:

Để chạy bất kỳ cell code nào sau đây, bạn có thể nhập **Shift + Enter** để tính toán code  trong một cell.

Cài đặt phiên bản Python.

In [1]:
# check Python version
!python -V

Python 3.10.14


Tải xuống thư viện và các tệp đi kèm mà chúng ta sẽ cần để chạy lab này.

In [2]:
import pandas as pd # import library to read data into dataframe
pd.set_option('display.max_columns', None)
import numpy as np # import numpy library
import re # import library for regular expression

Tải xuống dữ liệu từ máy chủ IBM và đọc nó vào khung dữ liệu (dataframe) *pandas*.

In [4]:
# Điền code vào đây
recipes = pd.read_csv("https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DS0103EN/labs/data/recipes.csv")

print("Data read into dataframe!") # takes about 30 seconds

Data read into dataframe!


Hiển thị một vài hàng đầu tiên.

In [5]:
recipes.head()

country almond angelica anise anise_seed apple apple_brandy apricot  \
0  Vietnamese     No       No    No         No    No           No      No   
1  Vietnamese     No       No    No         No    No           No      No   
2  Vietnamese     No       No    No         No    No           No      No   
3  Vietnamese     No       No    No         No    No           No      No   
4  Vietnamese     No       No    No         No    No           No      No   

  armagnac artemisia artichoke asparagus avocado bacon baked_potato balm  \
0       No        No        No        No      No    No           No   No   
1       No        No        No        No      No    No           No   No   
2       No        No        No        No      No    No           No   No   
3       No        No        No        No      No    No           No   No   
4       No        No        No        No      No    No           No   No   

  banana barley bartlett_pear basil bay bean beech beef beef_broth beef_liver  \
0     No     No            No   Yes  No   No    No   No         No         No   
1     No     No            No    No  No   No    No   No         No         No   
2     No     No            No    No  No   No    No   No         No         No   
3     No     No            No   Yes  No  Yes    No   No        Yes         No   
4     No     No            No    No  No   No    No   No         No         No   

  beer beet bell_pepper bergamot berry bitter_orange black_bean black_currant  \
0   No   No          No       No    No            No         No            No   
1   No   No          No       No    No            No         No            No   
2   No   No          No       No    No            No         No            No   
3   No   No          No       No    No            No         No            No   
4   No   No          No       No    No            No         No            No   

  black_mustard_seed_oil black_pepper black_raspberry black_sesame_seed  \
0                     No           No              No                No   
1                     No          Yes              No                No   
2                     No           No              No                No   
3                     No           No              No                No   
4                     No           No              No                No   

  black_tea blackberry blackberry_brandy blue_cheese blueberry bone_oil  \
0        No         No                No          No        No       No   
1        No         No                No          No        No       No   
2        No         No                No          No        No       No   
3        No         No                No          No        No       No   
4        No         No                No          No        No       No   

  bourbon_whiskey brandy brassica bread broccoli brown_rice brussels_sprout  \
0              No     No       No    No       No         No              No   
1              No     No       No    No       No         No              No   
2              No     No       No    No       No         No              No   
3              No     No       No    No       No         No              No   
4              No     No       No    No       No         No              No   

  buckwheat butter buttermilk cabbage cabernet_sauvignon_wine cacao  \
0        No     No         No      No                      No    No   
1        No     No         No      No                      No    No   
2        No     No         No      No                      No    No   
3        No     No         No      No                      No    No   
4        No     No         No      No                      No    No   

  camembert_cheese cane_molasses caraway cardamom carnation carob carrot  \
0               No            No      No       No        No    No    Yes   
1               No            No      No       No        No    No     No   
2               No            No      No       No        No    No     No   
3   

Nhấp đúp vào __đây__ để xem lời giải.
<!--
recipes.head()
-->

Lấy chiều từ khung dữ liệu.

In [6]:
recipes.shape

(57691, 384)

Vì vậy, tập dữ liệu của chúng ta gồm 57.691 công thức nấu ăn. Mỗi hàng đại diện cho một công thức và với mỗi công thức, món ăn tương ứng được ghi lại cũng như xem 384 thành phần nguyên liệu có tồn tại trong công thức hay không, bắt đầu bằng hạnh nhân (almond) và kết thúc bằng bí ngòi (zuchini).

Chúng ta biết rằng công thức sushi cơ bản bao gồm các thành phần:
* cơm
* xì dầu
* wasabi (mù tạt)
* một ít cá/rau

Hãy kiểm tra xem các thành phần này có tồn tại trong khung dữ liệu của chúng ta không:

In [7]:
ingredients = list(recipes.columns.values)

print([match.group(0) for ingredient in ingredients for match in [(re.compile(".*(rice).*")).search(ingredient)] if match])
print([match.group(0) for ingredient in ingredients for match in [(re.compile(".*(wasabi).*")).search(ingredient)] if match])
print([match.group(0) for ingredient in ingredients for match in [(re.compile(".*(soy).*")).search(ingredient)] if match])

['brown_rice', 'licorice', 'rice']
['wasabi']
['soy_sauce', 'soybean', 'soybean_oil']


Có các thành phần này!

* cơm tồn tại dưới dạng rice.
* mù tạt tồn tại dưới dạng wasabi.
* đậu nành tồn tại dưới dạng soy_sauce.

Vì vậy, có thể nếu một công thức có cả ba thành phần: cơm, mù tạt và đậu nành, thì chúng ta có thể tự tin nói rằng công thức đó là một món ăn **Nhật Bản**. Hãy ghi nhớ điều này!

----------------

# Chuẩn bị Dữ liệu <a id="4"></a>

<img src="https://ibm.box.com/shared/static/lqc2j3r0ndhokh77mubohwjqybzf8dhk.png" width=500>

Trong phần này, chúng ta sẽ chuẩn bị dữ liệu cho giai đoạn tiếp theo trong phương pháp luận khoa học dữ liệu, đó là mô hình hóa. Giai đoạn này bao gồm việc khám phá dữ liệu sâu hơn và đảm bảo rằng dữ liệu ở định dạng phù hợp cho thuật toán học máy mà chúng ta đã chọn trong giai đoạn tiếp cận phân tích, đó là cây quyết định.

Đầu tiên, hãy xem dữ liệu để xem liệu nó có cần làm sạch (cleaning) hay không.

In [8]:
# Điền code vào 
recipes["country"].value_counts() # frequency table

country
American        40150
Mexico           1754
Italian          1715
Italy            1461
Asian            1176
                ...  
Indonesia          12
Belgium            11
East-African       11
Israel              9
Bangladesh          4
Name: count, Length: 69, dtype: int64

Bằng cách nhìn vào bảng trên, chúng ta có thể đưa ra các nhận xét sau:

1. Cột Cuisine  (Món ăn) được gắn nhãn là Country (Quốc gia) là không chính xác.
2. Tên món ăn không nhất quán vì một số không viết hoa chữ cái đầu.
3. Một số món ăn bị trùng lặp dưới dạng biến tấu của tên quốc gia, chẳng hạn như Vietnam và Vietnamese.
4. Một số nền ẩm thực có rất ít công thức nấu ăn.

#### Hãy khắc phục những vấn đề này.

Sửa tên của cột hiển thị tên món ăn (cuisine)

In [9]:
column_names = recipes.columns.values

column_names[0] =  "cuisine"
recipes.columns = column_names

recipes

cuisine almond angelica anise anise_seed apple apple_brandy apricot  \
0      Vietnamese     No       No    No         No    No           No      No   
1      Vietnamese     No       No    No         No    No           No      No   
2      Vietnamese     No       No    No         No    No           No      No   
3      Vietnamese     No       No    No         No    No           No      No   
4      Vietnamese     No       No    No         No    No           No      No   
...           ...    ...      ...   ...        ...   ...          ...     ...   
57686       Japan     No       No    No         No    No           No      No   
57687       Japan     No       No    No         No    No           No      No   
57688       Japan     No       No    No         No    No           No      No   
57689       Japan     No       No    No         No    No           No      No   
57690       Japan     No       No    No         No    No           No      No   

      armagnac artemisia artichoke asparagus avocado bacon baked_potato balm  \
0           No        No        No        No      No    No           No   No   
1           No        No        No        No      No    No           No   No   
2           No        No        No        No      No    No           No   No   
3           No        No        No        No      No    No           No   No   
4           No        No        No        No      No    No           No   No   
...        ...       ...       ...       ...     ...   ...          ...  ...   
57686       No        No        No        No      No    No           No   No   
57687       No        No        No        No      No    No           No   No   
57688       No        No        No        No      No    No           No   No   
57689       No        No        No        No      No   Yes           No   No   
57690       No        No        No        No      No    No           No   No   

      banana barley bartlett_pear basil bay bean beech beef beef_broth  \
0         No     No            No   Yes  No   No    No   No         No   
1         No     No            No    No  No   No    No   No         No   
2         No     No            No    No  No   No    No   No         No   
3         No     No            No   Yes  No  Yes    No   No        Yes   
4         No     No            No    No  No   No    No   No         No   
...      ...    ...           ...   ...  ..  ...   ...  ...        ...   
57686     No     No            No    No  No   No    No   No         No   
57687     No     No            No    No  No   No    No   No         No   
57688     No    Yes            No    No  No   No    No   No         No   
57689     No     No            No    No  No   No    No   No         No   
57690     No    Yes            No    No  No   No    No   No         No   

      beef_liver beer beet bell_pepper bergamot berry bitter_orange  \
0             No   No   No          No       No    No            No   
1             No   No   No          No       No    No            No   
2             No   No   No          No       No    No            No   
3             No   No   No          No       No    No            No   
4             No   No   No          No       No    No            No   
...          ...  ...  ...         ...      ...   ...           ...   
57686         No   No   No          No       No    No            No   
57687         No   No   No          No       No    No            No   
57688         No   No   No          No       No    No            No   
57689         No   No   No          No       No    No            No   
57690         No   No   No          No       No    No            No   

      black_bean black_currant black_mustard_seed_oil black_pepper  \
0             No            No                     No           No   
1             No            No                     No          Yes   
2             No            No                     No           No   
3             No            No                     No 

Đảm bảo tên của các nền ẩm thực viết bằng chữ thường.

In [11]:
recipes["cuisine"] = recipes["cuisine"].str.lower()

recipes

cuisine almond angelica anise anise_seed apple apple_brandy apricot  \
0      vietnamese     No       No    No         No    No           No      No   
1      vietnamese     No       No    No         No    No           No      No   
2      vietnamese     No       No    No         No    No           No      No   
3      vietnamese     No       No    No         No    No           No      No   
4      vietnamese     No       No    No         No    No           No      No   
...           ...    ...      ...   ...        ...   ...          ...     ...   
57686       japan     No       No    No         No    No           No      No   
57687       japan     No       No    No         No    No           No      No   
57688       japan     No       No    No         No    No           No      No   
57689       japan     No       No    No         No    No           No      No   
57690       japan     No       No    No         No    No           No      No   

      armagnac artemisia artichoke asparagus avocado bacon baked_potato balm  \
0           No        No        No        No      No    No           No   No   
1           No        No        No        No      No    No           No   No   
2           No        No        No        No      No    No           No   No   
3           No        No        No        No      No    No           No   No   
4           No        No        No        No      No    No           No   No   
...        ...       ...       ...       ...     ...   ...          ...  ...   
57686       No        No        No        No      No    No           No   No   
57687       No        No        No        No      No    No           No   No   
57688       No        No        No        No      No    No           No   No   
57689       No        No        No        No      No   Yes           No   No   
57690       No        No        No        No      No    No           No   No   

      banana barley bartlett_pear basil bay bean beech beef beef_broth  \
0         No     No            No   Yes  No   No    No   No         No   
1         No     No            No    No  No   No    No   No         No   
2         No     No            No    No  No   No    No   No         No   
3         No     No            No   Yes  No  Yes    No   No        Yes   
4         No     No            No    No  No   No    No   No         No   
...      ...    ...           ...   ...  ..  ...   ...  ...        ...   
57686     No     No            No    No  No   No    No   No         No   
57687     No     No            No    No  No   No    No   No         No   
57688     No    Yes            No    No  No   No    No   No         No   
57689     No     No            No    No  No   No    No   No         No   
57690     No    Yes            No    No  No   No    No   No         No   

      beef_liver beer beet bell_pepper bergamot berry bitter_orange  \
0             No   No   No          No       No    No            No   
1             No   No   No          No       No    No            No   
2             No   No   No          No       No    No            No   
3             No   No   No          No       No    No            No   
4             No   No   No          No       No    No            No   
...          ...  ...  ...         ...      ...   ...           ...   
57686         No   No   No          No       No    No            No   
57687         No   No   No          No       No    No            No   
57688         No   No   No          No       No    No            No   
57689         No   No   No          No       No    No            No   
57690         No   No   No          No       No    No            No   

      black_bean black_currant black_mustard_seed_oil black_pepper  \
0             No            No                     No           No   
1             No            No                     No          Yes   
2             No            No                     No           No   
3             No            No                     No 

Để tên của các nền ẩm thực đồng nhất với nhau.

In [12]:
recipes.loc[recipes["cuisine"] == "austria", "cuisine"] = "austrian"
recipes.loc[recipes["cuisine"] == "belgium", "cuisine"] = "belgian"
recipes.loc[recipes["cuisine"] == "china", "cuisine"] = "chinese"
recipes.loc[recipes["cuisine"] == "canada", "cuisine"] = "canadian"
recipes.loc[recipes["cuisine"] == "netherlands", "cuisine"] = "dutch"
recipes.loc[recipes["cuisine"] == "france", "cuisine"] = "french"
recipes.loc[recipes["cuisine"] == "germany", "cuisine"] = "german"
recipes.loc[recipes["cuisine"] == "india", "cuisine"] = "indian"
recipes.loc[recipes["cuisine"] == "indonesia", "cuisine"] = "indonesian"
recipes.loc[recipes["cuisine"] == "iran", "cuisine"] = "iranian"
recipes.loc[recipes["cuisine"] == "italy", "cuisine"] = "italian"
recipes.loc[recipes["cuisine"] == "japan", "cuisine"] = "japanese"
recipes.loc[recipes["cuisine"] == "israel", "cuisine"] = "jewish"
recipes.loc[recipes["cuisine"] == "korea", "cuisine"] = "korean"
recipes.loc[recipes["cuisine"] == "lebanon", "cuisine"] = "lebanese"
recipes.loc[recipes["cuisine"] == "malaysia", "cuisine"] = "malaysian"
recipes.loc[recipes["cuisine"] == "mexico", "cuisine"] = "mexican"
recipes.loc[recipes["cuisine"] == "pakistan", "cuisine"] = "pakistani"
recipes.loc[recipes["cuisine"] == "philippines", "cuisine"] = "philippine"
recipes.loc[recipes["cuisine"] == "scandinavia", "cuisine"] = "scandinavian"
recipes.loc[recipes["cuisine"] == "spain", "cuisine"] = "spanish_portuguese"
recipes.loc[recipes["cuisine"] == "portugal", "cuisine"] = "spanish_portuguese"
recipes.loc[recipes["cuisine"] == "switzerland", "cuisine"] = "swiss"
recipes.loc[recipes["cuisine"] == "thailand", "cuisine"] = "thai"
recipes.loc[recipes["cuisine"] == "turkey", "cuisine"] = "turkish"
recipes.loc[recipes["cuisine"] == "vietnam", "cuisine"] = "vietnamese"
recipes.loc[recipes["cuisine"] == "uk-and-ireland", "cuisine"] = "uk-and-irish"
recipes.loc[recipes["cuisine"] == "irish", "cuisine"] = "uk-and-irish"

recipes

cuisine almond angelica anise anise_seed apple apple_brandy apricot  \
0      vietnamese     No       No    No         No    No           No      No   
1      vietnamese     No       No    No         No    No           No      No   
2      vietnamese     No       No    No         No    No           No      No   
3      vietnamese     No       No    No         No    No           No      No   
4      vietnamese     No       No    No         No    No           No      No   
...           ...    ...      ...   ...        ...   ...          ...     ...   
57686    japanese     No       No    No         No    No           No      No   
57687    japanese     No       No    No         No    No           No      No   
57688    japanese     No       No    No         No    No           No      No   
57689    japanese     No       No    No         No    No           No      No   
57690    japanese     No       No    No         No    No           No      No   

      armagnac artemisia artichoke asparagus avocado bacon baked_potato balm  \
0           No        No        No        No      No    No           No   No   
1           No        No        No        No      No    No           No   No   
2           No        No        No        No      No    No           No   No   
3           No        No        No        No      No    No           No   No   
4           No        No        No        No      No    No           No   No   
...        ...       ...       ...       ...     ...   ...          ...  ...   
57686       No        No        No        No      No    No           No   No   
57687       No        No        No        No      No    No           No   No   
57688       No        No        No        No      No    No           No   No   
57689       No        No        No        No      No   Yes           No   No   
57690       No        No        No        No      No    No           No   No   

      banana barley bartlett_pear basil bay bean beech beef beef_broth  \
0         No     No            No   Yes  No   No    No   No         No   
1         No     No            No    No  No   No    No   No         No   
2         No     No            No    No  No   No    No   No         No   
3         No     No            No   Yes  No  Yes    No   No        Yes   
4         No     No            No    No  No   No    No   No         No   
...      ...    ...           ...   ...  ..  ...   ...  ...        ...   
57686     No     No            No    No  No   No    No   No         No   
57687     No     No            No    No  No   No    No   No         No   
57688     No    Yes            No    No  No   No    No   No         No   
57689     No     No            No    No  No   No    No   No         No   
57690     No    Yes            No    No  No   No    No   No         No   

      beef_liver beer beet bell_pepper bergamot berry bitter_orange  \
0             No   No   No          No       No    No            No   
1             No   No   No          No       No    No            No   
2             No   No   No          No       No    No            No   
3             No   No   No          No       No    No            No   
4             No   No   No          No       No    No            No   
...          ...  ...  ...         ...      ...   ...           ...   
57686         No   No   No          No       No    No            No   
57687         No   No   No          No       No    No            No   
57688         No   No   No          No       No    No            No   
57689         No   No   No          No       No    No            No   
57690         No   No   No          No       No    No            No   

      black_bean black_currant black_mustard_seed_oil black_pepper  \
0             No            No                     No           No   
1             No            No                     No          Yes   
2             No            No                     No           No   
3             No            No                     No 

Loại bỏ các nền ẩm thực có < 50 công thức.

In [14]:
# get list of cuisines to keep
recipes_counts = recipes["cuisine"].value_counts()
cuisines_indices = recipes_counts > 50

# Điền code vào đây
cuisines_to_keep = list(np.array(recipes_counts.index.values)[np.array(cuisines_indices)])

In [15]:
rows_before = recipes.shape[0] # number of rows of original dataframe
print("Number of rows of original dataframe is {}.".format(rows_before))

# Điền code vào đây
recipes = recipes.loc[recipes['cuisine'].isin(cuisines_to_keep)]

rows_after = recipes.shape[0] # number of rows of processed dataframe
print("Number of rows of processed dataframe is {}.".format(rows_after))

print("{} rows removed!".format(rows_before - rows_after))

Number of rows of original dataframe is 57691.
Number of rows of processed dataframe is 57403.
288 rows removed!


Chuyển tất cả Yes thành 1 và No thành 0

In [ ]:
# Điền code vào đây
recipes = recipes.replace(...="Yes", value=...)
recipes = recipes....(to_replace=..., value=0)

#### Hãy phân tích dữ liệu thêm một chút để tìm hiểu dữ liệu tốt hơn và ghi nhận bất kỳ quan sát sơ bộ thú vị nào.

Chạy cell sau để nhận công thức có chứa **rice** *và* **soy** *và* **wasabi** *và* **seaweed**.

In [16]:
recipes.head()

cuisine almond angelica anise anise_seed apple apple_brandy apricot  \
0  vietnamese     No       No    No         No    No           No      No   
1  vietnamese     No       No    No         No    No           No      No   
2  vietnamese     No       No    No         No    No           No      No   
3  vietnamese     No       No    No         No    No           No      No   
4  vietnamese     No       No    No         No    No           No      No   

  armagnac artemisia artichoke asparagus avocado bacon baked_potato balm  \
0       No        No        No        No      No    No           No   No   
1       No        No        No        No      No    No           No   No   
2       No        No        No        No      No    No           No   No   
3       No        No        No        No      No    No           No   No   
4       No        No        No        No      No    No           No   No   

  banana barley bartlett_pear basil bay bean beech beef beef_broth beef_liver  \
0     No     No            No   Yes  No   No    No   No         No         No   
1     No     No            No    No  No   No    No   No         No         No   
2     No     No            No    No  No   No    No   No         No         No   
3     No     No            No   Yes  No  Yes    No   No        Yes         No   
4     No     No            No    No  No   No    No   No         No         No   

  beer beet bell_pepper bergamot berry bitter_orange black_bean black_currant  \
0   No   No          No       No    No            No         No            No   
1   No   No          No       No    No            No         No            No   
2   No   No          No       No    No            No         No            No   
3   No   No          No       No    No            No         No            No   
4   No   No          No       No    No            No         No            No   

  black_mustard_seed_oil black_pepper black_raspberry black_sesame_seed  \
0                     No           No              No                No   
1                     No          Yes              No                No   
2                     No           No              No                No   
3                     No           No              No                No   
4                     No           No              No                No   

  black_tea blackberry blackberry_brandy blue_cheese blueberry bone_oil  \
0        No         No                No          No        No       No   
1        No         No                No          No        No       No   
2        No         No                No          No        No       No   
3        No         No                No          No        No       No   
4        No         No                No          No        No       No   

  bourbon_whiskey brandy brassica bread broccoli brown_rice brussels_sprout  \
0              No     No       No    No       No         No              No   
1              No     No       No    No       No         No              No   
2              No     No       No    No       No         No              No   
3              No     No       No    No       No         No              No   
4              No     No       No    No       No         No              No   

  buckwheat butter buttermilk cabbage cabernet_sauvignon_wine cacao  \
0        No     No         No      No                      No    No   
1        No     No         No      No                      No    No   
2        No     No         No      No                      No    No   
3        No     No         No      No                      No    No   
4        No     No         No      No                      No    No   

  camembert_cheese cane_molasses caraway cardamom carnation carob carrot  \
0               No            No      No       No        No    No    Yes   
1               No            No      No       No        No    No     No   
2               No            No      No       No        No    No     No   
3   

In [17]:
check_recipes = recipes.loc[
    (recipes["rice"] == 1) &
    (recipes["soy_sauce"] == 1) &
    (recipes["wasabi"] == 1) &
    (recipes["seaweed"] == 1)
]

check_recipes

Empty DataFrame
Columns: [cuisine, almond, angelica, anise, anise_seed, apple, apple_brandy, apricot, armagnac, artemisia, artichoke, asparagus, avocado, bacon, baked_potato, balm, banana, barley, bartlett_pear, basil, bay, bean, beech, beef, beef_broth, beef_liver, beer, beet, bell_pepper, bergamot, berry, bitter_orange, black_bean, black_currant, black_mustard_seed_oil, black_pepper, black_raspberry, black_sesame_seed, black_tea, blackberry, blackberry_brandy, blue_cheese, blueberry, bone_oil, bourbon_whiskey, brandy, brassica, bread, broccoli, brown_rice, brussels_sprout, buckwheat, butter, buttermilk, cabbage, cabernet_sauvignon_wine, cacao, camembert_cheese, cane_molasses, caraway, cardamom, carnation, carob, carrot, cashew, cassava, catfish, cauliflower, caviar, cayenne, celery, celery_oil, cereal, chamomile, champagne_wine, chayote, cheddar_cheese, cheese, cherry, cherry_brandy, chervil, chicken, chicken_broth, chicken_liver, chickpea, chicory, chinese_cabbage, chive, cider, cilantro, cinnamon, citrus, citrus_peel, clam, clove, cocoa, coconut, coconut_oil, cod, coffee, ...]
Index: []

Dựa trên kết quả của code trên, chúng ta  có thể phân loại tất cả các công thức nấu ăn có chứa **rice** *và* **soy** *và* **wasabi** *và* **seaweed** thành  công thức nấu ăn của **Japanese** không? Tại sao?

Câu trả lời của bạn:

Nhấp đúp vào __đây__ để xem lời giải.
<!-- Câu trả lời đúng là:
Không, bởi vì các công thức nấu ăn khác như công thức nấu ăn Châu Á và Đông Á cũng chứa các thành phần này.
-->

Hãy đếm các thành phần trong tất cả các công thức nấu ăn.

In [18]:
# sum each column
# Điền code vào đây
ing = recipes.iloc[:, 1:].sum(axis=0)

In [ ]:
# define each column as a pandas series
# Điền code vào đây
ingredient = pd.Series(ing.index.values, index = np.arange(len(ing)))
count = pd.Series(list(ing), index = np.arange(len(ing)))

# create the dataframe
ing_df = pd.DataFrame(dict(ingredient = ingredient, count = count))
ing_df = ing_df[["ingredient", "count"]]
print(ing_df.to_string())

Bây giờ chúng ta có khung dữ liệu và tổng số các thành phần trên tất cả các công thức nấu ăn. Hãy sắp xếp khung dữ liệu này theo thứ tự giảm dần.

In [20]:
ing_df.sort_values(["count"], ascending=False, inplace=True)
ing_df.reset_index(inplace=True, drop=True)

print(ing_df)

      ingredient                                              count
0         garlic  YesYesYesNoYesYesYesYesNoYesNoYesYesYesYesYesY...
1           fish  YesYesNoYesYesYesYesYesYesYesNoYesYesYesNoYesY...
2        cayenne  YesYesNoYesYesNoYesYesYesYesNoYesYesNoNoYesYes...
3           seed  YesYesNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNo...
4     lime_juice  YesNoYesYesYesNoNoYesYesYesNoYesNoNoNoYesYesYe...
..           ...                                                ...
378     angelica  NoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNo...
379    carnation  NoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNo...
380  soybean_oil  NoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNo...
381      pimenta  NoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNo...
382       durian  NoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNoNo...

[383 rows x 2 columns]


#### 3 thành phần nguyên liệu phổ biến nhất là gì?

1. egg với 21.025 lần xuất hiện.
2. wheat với 20,781 lần xuất hiện.
3. butter với 20,719 lần xuất hiện.

Tuy nhiên, lưu ý rằng có một vấn đề với bảng trên. Có ~ 40.000 công thức nấu ăn của Mỹ trong tập dữ liệu, có nghĩa là dữ liệu thiên về các nguyên liệu của Mỹ.

**Do đó**, hãy tính toán một bản tóm tắt khách quan hơn về các thành phần bằng cách xem xét các thành phần của mỗi món ăn.

#### Hãy tạo một *profile* (hồ sơ) cho từng nền ẩm thực.

Nói cách khác, chúng ta hãy thử tìm hiểu xem người Trung Quốc thường sử dụng những nguyên liệu nào và ẩm thực **Canada** như thế nào chẳng hạn.

In [ ]:
cuisines = recipes.groupby("cuisine").mean()
cuisines.head()

Như được hiển thị ở trên, chúng ta vừa tạo khung dữ liệu trong đó mỗi hàng là một món ăn và mỗi cột (ngoại trừ cột đầu tiên) là một thành phần và các giá trị hàng thể hiện tỷ lệ phần trăm của mỗi thành phần trong món ăn tương ứng.

**Ví dụ**:

* *almond* có mặt trên 15,65% tất cả các công thức của **người châu Phi**.
* *butter* có mặt trên 38,11% tất cả các công thức của **người Canada**.

Hãy in hồ sơ cho mỗi món ăn bằng cách hiển thị 4 thành phần nguyên liệu hàng đầu trong từng nền ẩm thực.

In [ ]:
num_ingredients = 4 # define number of top ingredients to print

# define a function that prints the top ingredients for each cuisine
def print_top_ingredients(row):
    print(row.name.upper())
    row_sorted = row.sort_values(ascending=False)*100
    top_ingredients = list(row_sorted.index.values)[0:num_ingredients]
    row_sorted = list(row_sorted)[0:num_ingredients]

    for ind, ingredient in enumerate(top_ingredients):
        print("%s (%d%%)" % (ingredient, row_sorted[ind]), end=' ')
    print("\n")

# apply function to cuisines dataframe
create_cuisines_profiles = cuisines.apply(print_top_ingredients, axis=1)

Tại thời điểm này, chúng ta đã có thể hiểu rõ về dữ liệu và dữ liệu đã sẵn sàng và ở định dạng phù hợp để lập mô hình!

-----------

### Cảm ơn bạn đã hoàn thành lab này!

Nguồn: IBM